notebooks/implicit_differentiation.ipynb

```diff
@@ -0,0 +1,561 @@
+    "\n",
+    "    return state.position, get_maxgrad(state), num_iterations+1\n",
+    "\n",
+    "\n",
+    "def run_minimization_while_nl(\n",
+    "    neighbor_fn,\n",
+    "    energy_fn,\n",
+    "    R_init,\n",
+    "    forced_rebuilding,\n",
+    "    shift,\n",
+    "    max_grad_thresh=1e-12,\n",
+    "    max_num_steps=1000000,\n",
+    "    **kwargs\n",
+    "):\n",
+    "    init_fn, apply_fn = minimize.fire_descent(energy_fn, shift, **kwargs)\n",
+    "    apply_fn = jit(apply_fn)\n",
+    "\n",
+    "    nbrs = neighbor_fn.allocate(R_init)\n",
+    "    state = init_fn(R_init, neighbor=nbrs)\n",
+    "\n",
+    "    @jit\n",
+    "    def get_maxgrad(state):\n",
+    "        return jnp.amax(jnp.abs(state.force))\n",
+    "\n",
+    "    @jit\n",
+    "    def cond_fn(state, i):\n",
+    "        return jnp.logical_and(get_maxgrad(state) > max_grad_thresh, i < max_num_steps)\n",
+    "\n",
+    "    @jit\n",
+    "    def update_nbrs(R,nbrs):\n",
+    "        return neighbor_fn.update(R,nbrs)\n",
+    "    \n",
+    "    steps = 0\n",
+    "    while cond_fn(state, steps):\n",
+    "        nbrs = update_nbrs(state.position,nbrs)\n",
+    "        new_state = apply_fn(state, neighbor=nbrs)\n",
+    "        if forced_rebuilding and steps == 10:\n",
+    "            print(\"Forced rebuilding of neighbor_list.\")\n",
+    "            nbrs = neighbor_fn.allocate(state.position)\n",
+    "        if nbrs.did_buffer_overflow:\n",
+    "            print(\"Rebuilding neighbor_list.\")\n",
+    "            nbrs = neighbor_fn.allocate(state.position)\n",
+    "        else:\n",
+    "            state = new_state\n",
+    "            steps += 1\n",
+    "        \n",
+    "    return state.position, nbrs, steps+1\n",
+    "\n",
+    "# This version is fully differentiable and jit-able\n",
+    "def run_minimization_scan(force_fn, R_init, shift, num_steps=5000, **kwargs):\n",
+    "    init, apply = minimize.fire_descent(jit(force_fn), shift, **kwargs)\n",
+    "    apply = jit(apply)\n",
+    "\n",
+    "    @jit\n",
+    "    def scan_fn(state, i):\n",
+    "        return apply(state), 0.0\n",
+    "\n",
+    "    state = init(R_init)\n",
+    "    state, _ = lax.scan(scan_fn, state, jnp.arange(num_steps))\n",
+    "    return state.position, jnp.amax(jnp.abs(force_fn(state.position)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "introductory-priority",
+   "metadata": {},
+   "source": [
+    "# Meta Optimization"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "industrial-tonight",
+   "metadata": {},
+   "source": [
+    "In this notebook we'll have another look at differentiating through an energy minimization routine. But this time we'll look at how a technique called implicit differentiation lets us do so much more efficiently than before. Let us first set up our system and see for ourselves what goes wrong when we aren't using implicit differentiation."
```

Comment:
I know you cite it below, but it might be nice to mention that we're going to be using JaxOpt here.

Reply:
Done